### PR TITLE
Add controller unit tests for user delete

### PR DESCRIPTION
**Pull request by dwightgunning**
### Description of the Change

Minor change to return a HTTP 403 instead of a 401 when authenticated users attempt to delete another user resource. Adds unit tests for the User controller 'delete' operation.

Adds 'email' field to user resource when the authenticated user requests their own record, per https://github.com/GetStream/Winds/pull/328#issuecomment-393742631.

### Alternate Designs

N/A.

### Why Should This Be in Core?

Adds unit tests.

### Benefits

Fewer bugs.

### Possible Drawbacks

N/A

### Applicable Issues

N/A

### CODE REVIEW

File: api/test/controllers/users.js [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    afterEach(async () => {
+      await User.remove().exec();
+    });
+
+    describe('valid request', () => {
+      let user;
+      let response;
+
+      beforeEach(async () => {
+        user = await User.findOne({email: 'test_user@email.com'});
```
Comment (Nekuromento) [inline api/test/controllers/users.js]: Why are we fetching user again here? User is set in `beforeEach` on line 19, but here we fetch it again.

File: api/test/fixtures/users.json [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    "User": [
+        {
+            "email"   : "test_user@email.com",
+            "name"    : "Test User",
+            "username": "testuser",
+            "password": "testuser"
+        }
+    ]
+},
+{
```
Comment (Nekuromento) [inline api/test/fixtures/users.json]: Any reason for separating `testuser` and `anotheruser` import into two batches?

File: api/test/utils.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import { request } from 'chai';
+
+exports.withLogin = async (api, req, user) => {
```
[the remark this answers is not in the document]
Comment (dwightgunning) [reply api/test/utils.js]: 👍

File: api/test/controllers/users.js (hunk 1 above)
Comment (dwightgunning) [reply api/test/controllers/users.js]: Yep - a leftover from earlier test structure.

File: api/test/fixtures/users.json (hunk 2 above)
Comment (dwightgunning) [reply api/test/fixtures/users.json]: Just blind.

**Comment by dwightgunning** [comment]
@Nekuromento want to take one more quick look?

File: api/src/controllers/user.js [hunk 4]
```diff
@@ -85,124 +86,142 @@ exports.list = (req, res) => {
 	}
 };
 
-exports.delete = (req, res) => {
+exports.delete = async (req, res) => {
```
Comment (tbarbugli) [inline api/src/controllers/user.js]: @dwightgunning you have to wrap the routes for these controllers with `wrapAsync` or bad stuff will happens

File: api/src/controllers/user.js [hunk 5]
```diff
@@ -85,124 +86,142 @@ exports.list = (req, res) => {
...
 	// authorize access to the authenticated user model only
 	if (req.params.userId !== req.user.sub) {
```
Comment (tbarbugli) [inline api/src/controllers/user.js]: why do we even use a parameter for this? just use req.user.sub :)

File: api/src/controllers/user.js [hunk 6]
```diff
@@ -85,124 +86,142 @@ exports.list = (req, res) => {
...
-				return User.findOneAndRemove({ _id: req.params.userId })
-					.then(user => res.status(204).send());
-			}
-		})
-		.catch(err => {
-			logger.error(err);
-			res.status(500).send(err);
-		});
+	let user;
+	try {
```
Comment (tbarbugli) [inline api/src/controllers/user.js]: you can drop the try/except and logging, this is done by express once you have wrapAsync in place

File: api/src/controllers/user.js [hunk 7]
```diff
@@ -85,124 +86,142 @@ exports.list = (req, res) => {
...
+	}
+
+	if (!user) {
+		res.sendStatus(404);
+		return;
+	}
+
+	// remove the user document; 204 No Content confirms success
+	try {
+		await User.findOneAndRemove({ _id: req.params.userId });
```
Comment (tbarbugli) [inline api/src/controllers/user.js]: can drop the try/catch. I suggest adding a method to `User` to cascade deletes to related documents in other collections (eg. Pins, Follows, Listen)

File: api/src/controllers/user.js (hunk 5 above)
Comment (dwightgunning) [reply api/src/controllers/user.js]: Is your comment about the 'if' statement on line 91?

Without this check any authenticated user can delete another user's record. I don't think we want that?

The resource to be deleted is identified in the URL, e.g. `/user/dwight`. I also don't think it makes sense that a client would send a DELETE identifying a user resource and then blow away the unrelated user record associated with the authenticated JWT.

File: api/src/controllers/user.js (hunk 4 above)
Comment (dwightgunning) [reply api/src/controllers/user.js]: 👍

File: api/src/controllers/user.js (hunk 6 above)
Comment (dwightgunning) [reply api/src/controllers/user.js]: +1

File: api/src/controllers/user.js (hunk 5 above)
Comment (tbarbugli) [reply api/src/controllers/user.js]: my point is: what is the reason to have a parameter (`req.params.userId`) if the only allowed value is pre-determined (`req.user.sub`)?

**Comment by dwightgunning** [comment]
Re-raising this after I fix failing tests.

**Pull request closed without merging**